module/zfs/arc.c

```diff
@@ -8700,6 +9168,838 @@ l2arc_stop(void)
+	    L2ARC_DEV_HDR_EVICT_FIRST);
+
+	/* Prepare the rebuild processing state */
+	bcopy(l2dhdr->dh_start_lbps, lb_ptrs, sizeof (lb_ptrs));
+
+	/* Start the rebuild process */
+	for (int i = 0; i < l2dhdr->dh_log_blk_count; i++) {
+		if (!l2arc_log_blkptr_valid(dev, &lb_ptrs[0])) {
+			zfs_dbgmsg("L2ARC log block pointer invalid, "
+			    "offset: %llu, vdev guid: %llu, l2ad_start: %llu, "
+			    "l2ad_end: %llu, l2ad_hand: %llu, "
+			    "l2ad_evict: %llu, l2ad_first: %d, lbp_psize: %llu",
+			    lb_ptrs[0].lbp_daddr, dev->l2ad_vdev->vdev_guid,
+			    dev->l2ad_start, dev->l2ad_end, dev->l2ad_hand,
+			    dev->l2ad_evict, dev->l2ad_first,
+			    L2BLK_GET_PSIZE((&lb_ptrs[0])->lbp_prop));
+			err = SET_ERROR(EINVAL);
+			break;
+		}
+
+		if ((err = l2arc_log_blk_read(dev, &lb_ptrs[0], &lb_ptrs[1],
+		    this_lb, next_lb, this_io, &next_io)) != 0)
+			goto out;
+
+		/*
+		 * Our memory pressure valve. If the system is running low
+		 * on memory, rather than swamping memory with new ARC buf
+		 * hdrs, we opt not to rebuild the L2ARC. At this point,
+		 * however, we have already set up our L2ARC dev to chain in
+		 * new metadata log blocks, so the user may choose to offline/
+		 * online the L2ARC dev at a later time (or re-import the pool)
+		 * to reconstruct it (when there's less memory pressure).
+		 */
+		if (arc_reclaim_needed()) {
+			ARCSTAT_BUMP(arcstat_l2_rebuild_abort_lowmem);
+			cmn_err(CE_NOTE, "System running low on memory, "
+			    "aborting L2ARC rebuild.");
+			err = SET_ERROR(ENOMEM);
+			goto out;
+		}
+
+		/*
+		 * We have to drop the lock after checking for memory pressure,
+		 * otherwise we goto out and later call l2arc_dev_hdr_update()
+		 * which panics in zio.c because we dropped the lock
+		 * prematurely.
+		 */
+		spa_config_exit(spa, SCL_L2ARC, vd);
+		lock_held = B_FALSE;
+
+		/*
+		 * Now that we know that the next_lb checks out alright, we
+		 * can start reconstruction from this log block.
+		 */
+		l2arc_log_blk_restore(dev, this_lb,
+		    L2BLK_GET_PSIZE((&lb_ptrs[0])->lbp_prop),
+		    lb_ptrs[0].lbp_daddr);
+
+		/*
+		 * log block restored, include its pointer in the list of
+		 * pointers to log blocks present in the L2ARC device.
+		 */
+		lb_ptr_buf = kmem_zalloc(sizeof (l2arc_lb_ptr_buf_t), KM_SLEEP);
+		lb_ptr_buf->lb_ptr = kmem_zalloc(sizeof (l2arc_log_blkptr_t),
+		    KM_SLEEP);
+		bcopy(&lb_ptrs[0], lb_ptr_buf->lb_ptr,
+		    sizeof (l2arc_log_blkptr_t));
+		mutex_enter(&dev->l2ad_mtx);
+		list_insert_tail(&dev->l2ad_lbptr_list, lb_ptr_buf);
+		mutex_exit(&dev->l2ad_mtx);
+		zfs_refcount_add(&dev->l2ad_log_blk_count, lb_ptr_buf);
+		vdev_space_update(vd,
+		    L2BLK_GET_PSIZE((&lb_ptrs[0])->lbp_prop), 0, 0);
+
+		for (;;) {
+			if (dev->l2ad_rebuild_cancel) {
+				mutex_enter(&l2arc_rebuild_thr_lock);
+				dev->l2ad_rebuild = B_FALSE;
+				cv_signal(&l2arc_rebuild_thr_cv);
+				mutex_exit(&l2arc_rebuild_thr_lock);
+				err = SET_ERROR(ECANCELED);
+				goto out;
```

Comment:
If the first time through this loop `l2ad_rebuild_cancel` is true then we'll goto out but will panic (based on the comment at line 9303) because we won't be holding the `SCL_L2ARC` lock.

Reply:
Good point, will take a look. In any case we want to avoid updating the header if the rebuild is cancelled because of `l2arc_remove_vdev()`. Will fix.